File: 2023/java/D01P2.java
import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.IOException;

public class d01p2 {
    private static final int ZERO = 48;
    private static final int NINE = 57;

    public static void main(String[] args) throws IOException {
        System.out.println("Day 1 - Part 2");

        BufferedReader reader = new BufferedReader(new FileReader(new File("../_input/day01.txt")));
        
        int total = 0;

        String line;
        while((line = reader.readLine()) != null) {
            total += get(line.trim());
        }

        System.out.println("Total: " + total);
    }

    private static int get(String str) {
        for(int i = 0; i < str.length(); i++) {
            if(str.startsWith("one", i)) {
                str = str.substring(0, i) + "1" + str.substring(i + 2);
            }
            else if(str.startsWith("two", i)) {
                str = str.substring(0, i) + "2" + str.substring(i + 2);
            }
            else if(str.startsWith("three", i)) {
                str = str.substring(0, i) + "3" + str.substring(i + 4);
            }
            else if(str.startsWith("four", i)) {
                str = str.substring(0, i) + "4" + str.substring(i + 3);
            }
            else if(str.startsWith("five", i)) {
                str = str.substring(0, i) + "5" + str.substring(i + 3);
            }
            else if(str.startsWith("six", i)) {
                str = str.substring(0, i) + "6" + str.substring(i + 2);
            }
            else if(str.startsWith("seven", i)) {
                str = str.substring(0, i) + "7" + str.substring(i + 4);
            }
            else if(str.startsWith("eight", i)) {
                str = str.substring(0, i) + "8" + str.substring(i + 4);
            }
            else if(str.startsWith("nine", i)) {
                str = str.substring(0, i) + "9" + str.substring(i + 3);
            }
        }

        int first = -1;
        int last = -1;

        for(int c : str.chars().toArray()) {
            if(c >= ZERO && c <= NINE) {
                if(first == -1) first = c - ZERO;
                else {
                    last = c - ZERO;
                }
            }
        }

        if(last == -1) last = first;
        return (first * 10) + last;
    }
}

File: 2023/rust/src/bin/d01p2.rs
use std::fs;

fn main() {
    println!("Day 01 - Part 2");

    let str = fs::read_to_string("../_input/day1.txt").unwrap();
    let lines = str.lines();

    let mut total = 0;
    for l in lines {
        total += get(l);
    }

    println!("Total: {}", total);
}

fn get(str: &str) -> u32 {
    let mut str = String::from(str);
    let mut i = 0_usize;
    while i < str.len() {
        let mut replace = ("0", 0);
        {
            let sub = &str[i..str.len()];
            if sub.starts_with("one") {
                replace = ("1", 2);
            } else if sub.starts_with("two") {
                replace = ("2", 2);
            } else if sub.starts_with("three") {
                replace = ("3", 4);
            } else if sub.starts_with("four") {
                replace = ("4", 3);
            } else if sub.starts_with("five") {
                replace = ("5", 3);
            } else if sub.starts_with("six") {
                replace = ("6", 2);
            } else if sub.starts_with("seven") {
                replace = ("7", 4);
            } else if sub.starts_with("eight") {
                replace = ("8", 4);
            } else if sub.starts_with("nine") {
                replace = ("9", 3);
            }
        }

        if replace.1 != 0 {
            str.replace_range(i..(i + replace.1), replace.0);
        }

        i += 1;
    }

    let str = str.as_str();

    let mut first = None;
    let mut last = None;

    for c in str.chars() {
        if c.is_ascii_digit() {
            if first.is_none() {
                first = c.to_digit(10);
            } else {
                last = c.to_digit(10);
            }
        }
    }

    if last.is_none() {
        last = first;
    }

    (first.unwrap() * 10) + last.unwrap()
}

#[cfg(test)]
mod test {
    use crate::get;

    #[test]
    fn test() {
        assert_eq!(get("two1nine"), 29);
        assert_eq!(get("eightwothree"), 83);
        assert_eq!(get("abcone2threexyz"), 13);
        assert_eq!(get("xtwone3four"), 24);
        assert_eq!(get("4nineeightseven2"), 42);
        assert_eq!(get("zoneight234"), 14);
        assert_eq!(get("7pqrstsixteen"), 76);

        assert_eq!(get("oneight"), 18);
    }
}
